oai-pmh-handler/src/main/java/no/sikt/nva/oai/pmh/handler/DefaultOaiPmhDataProvider.java

```diff
@@ -116,26 +71,66 @@ private JAXBElement<OAIPMHtype> baseResponse() {
   }
 
   private JAXBElement<OAIPMHtype> listSets() {
-    var response = baseResponse();
-    var value = response.getValue();
+    final ResourceSearchQuery query = buildAllAggregationsQuery();
+
+    var instanceTypes = doSearchAndExtractInstanceTypesFromTypeAggregation(query);
+
+    var uaiResponse = baseResponse();
+    var value = uaiResponse.getValue();
     value.getRequest().setVerb(VerbType.LIST_SETS);
 
     var listSets = objectFactory.createListSetsType();
-    listSets.getSet().addAll(generateSets(objectFactory));
+    listSets.getSet().addAll(generateSets(objectFactory, instanceTypes));
     value.setListSets(listSets);
 
-    return response;
+    return uaiResponse;
+  }
+
+  private Set<String> doSearchAndExtractInstanceTypesFromTypeAggregation(
+      ResourceSearchQuery query) {
+    final HttpResponseFormatter<ResourceParameter> response;
+    try {
+      response = query.doSearch(opensearchClient);
+    } catch (RuntimeException e) {
+      LOGGER.error("Failed to search for publication instance types using 'type' aggregation.", e);
+      throw new ResourceSearchException("Error looking up instance types using aggregations.", e);
+    }
+
+    return response
+        .toPagedResponse()
+        .aggregations()
+        .getOrDefault(INSTANCE_TYPE_AGGREGATION_NAME, Collections.emptyList())
+        .stream()
+        .map(Facet::key)
+        .collect(Collectors.toSet());
+  }
+
+  private static ResourceSearchQuery buildAllAggregationsQuery() {
+    final ResourceSearchQuery query;
+    try {
+      query =
+          ResourceSearchQuery.builder()
+              .withParameter(ResourceParameter.AGGREGATION, AGGREGATION_ALL)
+              .withParameter(ResourceParameter.FROM, ZERO)
+              .withParameter(ResourceParameter.SIZE, ZERO)
+              .build();
+    } catch (BadRequestException e) {
+      // should never happen unless query validation code is changed!
+      LOGGER.error("Failed to search for publication instance types using 'type' aggregation.", e);
+      throw new ResourceSearchException("Error looking up instance types using aggregations.", e);
```

Comment:
In order to make this testable, can we inject the search query builder?

Reply:
I will investigate that in the next PR.